File: windows/src/BentleyRemote.Agent/Security/DesktopPreviewCrypto.cs
using System.Security.Cryptography;
using System.Text;

namespace BentleyRemote.Agent.Security;

/// <summary>Domain-separated authenticated encryption for one paired-device desktop preview.</summary>
internal static class DesktopPreviewCrypto
{
    private static readonly byte[] Salt = new byte[32];
    private static readonly byte[] Info = Encoding.UTF8.GetBytes("bentley-remote/v1/desktop-preview/aes-256-gcm");

    public static (string Nonce, string Ciphertext) Encrypt(byte[] pairingSecret, string requestId,
        long capturedAt, string mimeType, byte[] image)
    {
        var key = PairingCrypto.Hkdf(pairingSecret, Salt, Info, 32);
        var nonce = RandomNumberGenerator.GetBytes(12);
        var ciphertext = new byte[image.Length + 16];
        try
        {
            using var aes = new AesGcm(key, 16);
            aes.Encrypt(nonce, image, ciphertext.AsSpan(0, image.Length), ciphertext.AsSpan(image.Length),
                Encoding.UTF8.GetBytes(Aad(requestId, capturedAt, mimeType)));
            return (Convert.ToBase64String(nonce), Convert.ToBase64String(ciphertext));
        }
        finally { CryptographicOperations.ZeroMemory(key); }
    }

    public static string Aad(string requestId, long capturedAt, string mimeType) =>
        $"bentley-remote/v1/desktop-preview\n{requestId}\n{capturedAt}\n{mimeType}";
}

File: android/app/src/main/java/com/bentley/remote/security/DesktopPreviewCrypto.kt
package com.bentley.remote.security

import android.util.Base64
import java.nio.charset.StandardCharsets
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/** Decrypts an image only after the ordinary paired WebSocket authentication has completed. */
object DesktopPreviewCrypto {
    private val salt = ByteArray(32)
    private val info = "bentley-remote/v1/desktop-preview/aes-256-gcm".toByteArray(StandardCharsets.UTF_8)

    fun decrypt(secret: ByteArray, requestId: String, capturedAt: Long, mimeType: String,
        nonceBase64: String, ciphertextBase64: String): ByteArray {
        val nonce = Base64.decode(nonceBase64, Base64.NO_WRAP)
        val ciphertext = Base64.decode(ciphertextBase64, Base64.NO_WRAP)
        require(nonce.size == 12) { "invalid desktop preview nonce" }
        require(ciphertext.size in 17..(1_048_576 + 16)) { "desktop preview exceeds size limit" }
        val key = hkdf(secret, salt, info, 32)
        return try {
            Cipher.getInstance("AES/GCM/NoPadding").run {
                init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
                updateAAD(aad(requestId, capturedAt, mimeType).toByteArray(StandardCharsets.UTF_8))
                doFinal(ciphertext)
            }
        } finally { key.fill(0) }
    }

    fun aad(requestId: String, capturedAt: Long, mimeType: String) =
        "bentley-remote/v1/desktop-preview\n$requestId\n$capturedAt\n$mimeType"

    private fun hkdf(input: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        val prk = PairingCrypto.hmac(salt, input)
        val output = ByteArray(length)
        var previous = byteArrayOf()
        var offset = 0
        var counter = 1
        while (offset < length) {
            previous = PairingCrypto.hmac(prk, previous + info + byteArrayOf(counter++.toByte()))
            val take = minOf(previous.size, length - offset)
            previous.copyInto(output, offset, 0, take)
            offset += take
        }
        prk.fill(0)
        return output
    }
}
